File: src/food/food.cpp
#include "food/food.h"
#include "health/health.h"
#include "locale/locale_manager.h"

#include <algorithm>
#include <sstream>

Food::Food(const std::string &name,
           const std::vector<Health> &health,
           int price,
           FoodType type,
           int initialFreshness)
    : name(name),
      baseHealth(health),
      price(price),
      type(type),
      freshness(initialFreshness),
      maxFreshness(initialFreshness)
{
}

void Food::UpdateFreshness(int gameMinutes)
{
    // 将分钟转换为小时
    float hours = static_cast<float>(gameMinutes) / 60.0f;
    float decay = GetDecayRate() * hours;
    freshness -= static_cast<int>(decay);
    // 不设下限，允许负数表示腐烂程度
}

std::string Food::GetFreshnessLabel() const
{
    LocaleManager *locale = LocaleManager::GetInstance();
    if (freshness > 75)
    {
        return locale->Get("food.freshness", "fresh");
    }
    else if (freshness > 50)
    {
        return locale->Get("food.freshness", "fairly_fresh");
    }
    else if (freshness > 25)
    {
        return locale->Get("food.freshness", "not_fresh");
    }
    else if (freshness > 0)
    {
        return locale->Get("food.freshness", "expiring");
    }
    else
    {
        return locale->Get("food.freshness", "expired");
    }
}

float Food::GetNutritionMultiplier() const
{
    if (freshness > 75)
    {
        return 1.0f;  // 100% 营养
    }
    else if (freshness > 50)
    {
        return 0.8f;  // 80% 营养
    }
    else if (freshness > 25)
    {
        return 0.5f;  // 50% 营养
    }
    else if (freshness > 0)
    {
        return 0.2f;  // 20% 营养
    }
    else
    {
        return -0.5f; // 过期：负效果
    }
}

std::vector<Health> Food::GetEffectiveHealth() const
{
    std::vector<Health> result;
    float multiplier = GetNutritionMultiplier();

    if (multiplier > 0)
    {
        // 正常效果：按比例计算营养
        for (const auto &h : baseHealth)
        {
            int effectiveValue = static_cast<int>(h.GetReserves() * multiplier);
            if (effectiveValue > 0)
            {
                result.push_back(Health(h.GetInfo(), effectiveValue));
            }
        }
    }
    else
    {
        // 过期食物：减少饱腹感和幸福感，其他营养为0
        result.push_back(Health("饱腹感", -10));
        result.push_back(Health("幸福感", -15));
        result.push_back(Health("饥渴", -5));
    }

    return result;
}

float Food::GetFoodPoisoningRisk() const
{
    if (freshness > 25)
    {
        return 0.0f;  // 不会食物中毒
    }
    else if (freshness > 0)
    {
        return 0.1f;  // 10% 概率
    }
    else if (freshness > -25)
    {
        return 0.3f;  // 30% 概率
    }
    else if (freshness > -50)
    {
        return 0.6f;  // 60% 概率
    }
    else
    {
        return 0.9f;  // 90% 概率
    }
}

std::string Food::GetInfo() const
{
    std::ostringstream ss;
    ss << name << " " << price << "元";

    // 添加新鲜度标签
    std::string label = GetFreshnessLabel();
    if (label != "新鲜")
    {
        ss << " [" << label << "]";
    }

    return ss.str();
}

float Food::GetDecayRate() const
{
    switch (type)
    {
    case FoodType::Fresh:
        return 4.0f;  // 25小时完全腐烂
    case FoodType::Beverage:
        return 2.0f;  // 50小时
    case FoodType::Processed:
        return 1.0f;  // 100小时
    case FoodType::Canned:
        return 0.1f;  // 1000小时
    default:
        return 1.0f;
    }
}

std::string FoodTypeToString(FoodType type)
{
    LocaleManager *locale = LocaleManager::GetInstance();
    switch (type)
    {
    case FoodType::Fresh:
        return locale->Get("food.type", "fresh");
    case FoodType::Beverage:
        return locale->Get("food.type", "beverage");
    case FoodType::Processed:
        return locale->Get("food.type", "processed");
    case FoodType::Canned:
        return locale->Get("food.type", "canned");
    default:
        return locale->Get("food.type", "unknown");
    }
}

File: internal/food/food.go
// Package food 定义食物类型和属性
package food

// FoodType 食物类型
type FoodType int

const (
	FoodTypeFresh     FoodType = iota // 新鲜食品
	FoodTypeBeverage                  // 饮料
	FoodTypeProcessed                 // 加工食品
	FoodTypeCanned                    // 罐头食品
)

// DecayRates 衰减率（每小时）
var DecayRates = map[FoodType]float64{
	FoodTypeFresh:     4.0, // 25小时完全腐烂
	FoodTypeBeverage:  2.0, // 50小时
	FoodTypeProcessed: 1.0, // 100小时
	FoodTypeCanned:    0.1, // 1000小时
}

// FoodTypeNames 食物类型名称
var FoodTypeNames = map[FoodType]string{
	FoodTypeFresh:     "新鲜",
	FoodTypeBeverage:  "饮料",
	FoodTypeProcessed: "加工",
	FoodTypeCanned:    "罐头",
}

// Health 营养效果
type Health struct {
	Name   string // 营养名称
	Amount int    // 营养值
}

// Food 食物定义
type Food struct {
	Name         string
	BaseHealth   []Health // 基础营养效果
	Price        int
	Type         FoodType
	Freshness    int // 当前新鲜度（0-100，可为负）
	MaxFreshness int // 最大新鲜度
}

// NewFood 创建新食物
func NewFood(name string, price int, foodType FoodType, health []Health) *Food {
	return &Food{
		Name:         name,
		BaseHealth:   health,
		Price:        price,
		Type:         foodType,
		Freshness:    100,
		MaxFreshness: 100,
	}
}

// UpdateFreshness 更新新鲜度（基于游戏时间）
func (f *Food) UpdateFreshness(gameMinutes int) {
	hours := float64(gameMinutes) / 60.0
	decay := DecayRates[f.Type] * hours
	f.Freshness -= int(decay)
}

// GetFreshnessLabel 获取新鲜度标签
func (f *Food) GetFreshnessLabel() string {
	if f.Freshness > 75 {
		return "新鲜"
	} else if f.Freshness > 50 {
		return "较新鲜"
	} else if f.Freshness > 25 {
		return "不太新鲜"
	} else if f.Freshness > 0 {
		return "即将过期"
	}
	return "已过期"
}

// GetNutritionMultiplier 获取营养倍率
func (f *Food) GetNutritionMultiplier() float64 {
	if f.Freshness > 75 {
		return 1.0 // 100%
	} else if f.Freshness > 50 {
		return 0.8 // 80%
	} else if f.Freshness > 25 {
		return 0.5 // 50%
	} else if f.Freshness > 0 {
		return 0.2 // 20%
	}
	return -0.5 // 过期：负效果
}

// GetEffectiveHealth 获取实际营养效果
func (f *Food) GetEffectiveHealth() []Health {
	multiplier := f.GetNutritionMultiplier()

	if multiplier > 0 {
		result := make([]Health, 0, len(f.BaseHealth))
		for _, h := range f.BaseHealth {
			effectiveValue := int(float64(h.Amount) * multiplier)
			if effectiveValue > 0 {
				result = append(result, Health{Name: h.Name, Amount: effectiveValue})
			}
		}
		return result
	}

	// 过期食物：减少饱腹感和幸福感
	return []Health{
		{Name: "饱腹感", Amount: -10},
		{Name: "幸福感", Amount: -15},
		{Name: "饥渴", Amount: -5},
	}
}

// GetFoodPoisoningRisk 获取食物中毒风险
func (f *Food) GetFoodPoisoningRisk() float64 {
	if f.Freshness > 25 {
		return 0.0
	} else if f.Freshness > 0 {
		return 0.1
	} else if f.Freshness > -25 {
		return 0.3
	} else if f.Freshness > -50 {
		return 0.6
	}
	return 0.9
}

// GetInfo 获取显示信息
func (f *Food) GetInfo() string {
	info := f.Name + " " + formatPrice(f.Price)
	label := f.GetFreshnessLabel()
	if label != "新鲜" {
		info += " [" + label + "]"
	}
	return info
}

func formatPrice(price int) string {
	return intToString(price) + "元"
}

func intToString(n int) string {
	if n == 0 {
		return "0"
	}
	result := ""
	negative := n < 0
	if negative {
		n = -n
	}
	for n > 0 {
		result = string(rune('0'+n%10)) + result
		n /= 10
	}
	if negative {
		result = "-" + result
	}
	return result
}
